File: frontend/de/unika/ipd/grgen/ast/expr/graph/VisitedNode.java
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 8.0
 * Copyright (C) 2003-2025 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos; and free programmers
 * licensed under LGPL v3 (see LICENSE.txt included in the packaging of this file)
 * www.grgen.de / www.grgen.net
 */

package de.unika.ipd.grgen.ast.expr.graph;

import java.util.Collection;
import java.util.Vector;

import de.unika.ipd.grgen.ast.*;
import de.unika.ipd.grgen.ast.expr.ExprNode;
import de.unika.ipd.grgen.ast.model.type.EdgeTypeNode;
import de.unika.ipd.grgen.ast.model.type.NodeTypeNode;
import de.unika.ipd.grgen.ast.type.TypeNode;
import de.unika.ipd.grgen.ast.type.basic.BasicTypeNode;
import de.unika.ipd.grgen.ast.type.basic.UntypedExecVarTypeNode;
import de.unika.ipd.grgen.ir.IR;
import de.unika.ipd.grgen.ir.expr.Expression;
import de.unika.ipd.grgen.ir.expr.graph.Visited;
import de.unika.ipd.grgen.parser.Coords;

public class VisitedNode extends ExprNode
{
	static {
		setName(VisitedNode.class, "visited");
	}

	private ExprNode visitorIDExpr;
	private ExprNode entityExpr;

	public VisitedNode(Coords coords, ExprNode visitorIDExpr, ExprNode entityExpr)
	{
		super(coords);

		this.visitorIDExpr = visitorIDExpr;
		becomeParent(visitorIDExpr);

		this.entityExpr = entityExpr;
		becomeParent(entityExpr);
	}

	@Override
	public Collection<? extends BaseNode> getChildren()
	{
		Vector<BaseNode> children = new Vector<BaseNode>();
		children.add(visitorIDExpr);
		children.add(entityExpr);
		return children;
	}

	@Override
	public Collection<String> getChildrenNames()
	{
		Vector<String> childrenNames = new Vector<String>();
		childrenNames.add("visitorID");
		childrenNames.add("entity");
		return childrenNames;
	}

	@Override
	protected boolean resolveLocal()
	{
		return true;
	}

	@Override
	protected boolean checkLocal()
	{
		if(visitorIDExpr.getType() instanceof UntypedExecVarTypeNode) {
			return true;
		}
		if(!visitorIDExpr.getType().isEqual(BasicTypeNode.intType)) {
			visitorIDExpr.reportError("The visited construct expects as index argument (visitorId) a value of type int"
					+ " (but is given a value of type " + visitorIDExpr.getType().getTypeName() + ").");
			return false;
		}
		if(entityExpr.getType() instanceof UntypedExecVarTypeNode) {
			return true;
		}
		if(entityExpr.getType() instanceof EdgeTypeNode) {
			return true;
		}
		if(entityExpr.getType() instanceof NodeTypeNode) {
			return true;
		}
		reportError("The visited construct expects as entity argument a value of type node or edge"
				+ " (but is given a value of type " + entityExpr.getType().getTypeName() + ").");
		return true;
	}

	@Override
	protected IR constructIR()
	{
		visitorIDExpr = visitorIDExpr.evaluate();
		entityExpr = entityExpr.evaluate();
		return new Visited(visitorIDExpr.checkIR(Expression.class), entityExpr.checkIR(Expression.class));
	}

	@Override
	public TypeNode getType()
	{
		return BasicTypeNode.booleanType;
	}
}

File: frontend/de/unika/ipd/grgen/ast/expr/graph/VisitedNode.cs
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 8.1
 * Copyright (C) 2003-2026 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos; and free programmers
 * licensed under LGPL v3, some components/parts use different licenses (see LICENSE.txt included in the packaging of this file)
 * www.grgen.de / www.grgen.net
 */

namespace de.unika.ipd.grgen.ast.expr.graph
{

	using System.Collections.Generic;

	using de.unika.ipd.grgen.ast;
	using ExprNode = de.unika.ipd.grgen.ast.expr.ExprNode;
	using EdgeTypeNode = de.unika.ipd.grgen.ast.model.type.EdgeTypeNode;
	using NodeTypeNode = de.unika.ipd.grgen.ast.model.type.NodeTypeNode;
	using TypeNode = de.unika.ipd.grgen.ast.type.TypeNode;
	using BasicTypeNode = de.unika.ipd.grgen.ast.type.basic.BasicTypeNode;
	using UntypedExecVarTypeNode = de.unika.ipd.grgen.ast.type.basic.UntypedExecVarTypeNode;
	using IR = de.unika.ipd.grgen.ir.IR;
	using Expression = de.unika.ipd.grgen.ir.expr.Expression;
	using Visited = de.unika.ipd.grgen.ir.expr.graph.Visited;
	using Coords = de.unika.ipd.grgen.parser.Coords;

	public class VisitedNode : ExprNode
	{
		static VisitedNode()
		{
			SetClassName(typeof(VisitedNode), "visited");
		}

		private ExprNode visitorIDExpr;
		private ExprNode entityExpr;

		public VisitedNode(Coords coords, ExprNode visitorIDExpr, ExprNode entityExpr)
			: base(coords)
		{

			this.visitorIDExpr = visitorIDExpr;
			BecomeParent(visitorIDExpr);

			this.entityExpr = entityExpr;
			BecomeParent(entityExpr);
		}

		public override ICollection<BaseNode> Children
		{
			get
			{
				IList<BaseNode> children = new List<BaseNode>();
				children.Add(visitorIDExpr);
				children.Add(entityExpr);
				return children;
			}
		}

		public override ICollection<string> ChildrenNames
		{
			get
			{
				IList<string> childrenNames = new List<string>();
				childrenNames.Add("visitorID");
				childrenNames.Add("entity");
				return childrenNames;
			}
		}

		protected internal override bool ResolveLocal()
		{
			return true;
		}

		protected internal override bool CheckLocal()
		{
			if(visitorIDExpr.Type is UntypedExecVarTypeNode)
				return true;
			if(!visitorIDExpr.Type.IsEqual(BasicTypeNode.intType))
			{
				visitorIDExpr.ReportError("The visited construct expects as index argument (visitorId) a value of type int"
						+ " (but is given a value of type " + visitorIDExpr.Type.TypeName + ").");
				return false;
			}
			if(entityExpr.Type is UntypedExecVarTypeNode)
				return true;
			if(entityExpr.Type is EdgeTypeNode)
				return true;
			if(entityExpr.Type is NodeTypeNode)
				return true;
			ReportError("The visited construct expects as entity argument a value of type node or edge"
					+ " (but is given a value of type " + entityExpr.Type.TypeName + ").");
			return true;
		}

		protected internal override IR ConstructIR()
		{
			visitorIDExpr = visitorIDExpr.Evaluate();
			entityExpr = entityExpr.Evaluate();
			return new Visited(visitorIDExpr.CheckIR<Expression>(typeof(Expression)), entityExpr.CheckIR<Expression>(typeof(Expression)));
		}

		public override TypeNode Type
		{
			get
			{
				return BasicTypeNode.booleanType;
			}
		}
	}

}
